0011-command-execution-protocol/README.md

```diff
@@ -0,0 +1,170 @@
+
+## UC3: OTA update
+> Note: while OTA updates require more support and probably a separate protocol, a basic version can be implemented using command execution.
+
+An endpoint may receive a command to update its firmware. When update is complete, the endpoint sends an update result to the server.
+
+# Design
+The Command Execution Protocol extension consists of two resource sets with the corresponding resource paths:
+- `/<endpoint_token>/command/<command_type>` - command resource
+- `/<endpoint_token>/result/<command_type>` - result resource
+
+## Command resource
+Command resource is used to distribute commands to the endpoints.
+
+The command resource is a request/response resource with an observe extension which resides at the following resource path:
+```
+/<endpoint_token>/command/<command_type>
+```
+where `<command_type>` is a type of the command the endpoint wants to serve.
+
+Command type MUST be a non-empty alpha-numeric string identifying the command type to the endpoint.
+
+### Request
+The endpoint requests MUST be a UTF-8 encoded JSON object with the JSON Schema defined in [command-request.schema.json](./command-request.schema.json) file.
+
+```json
+{
+  "$schema": "http://json-schema.org/schema#",
+  "title": "11/CEP command request schema",
+
+  "type": "object",
+  "properties": {
+    "observe": {
+      "type": "boolean",
+      "description": "Whether to send upcoming command for this endpoint."
+    }
+  },
+  "additionalProperties": false
+}
+```
+
+If `observe` field is present and is `true`, the server MUST send new commands to the endpoint whenever they arrive.
+
+If `observe` field is present and is `false`, the server MUST NOT send new commands to the endpoint whenever they arrive.
+
+The server MAY send commands to the endpoint when no request were done before, or `observe` was not present in a request.
+
+The server MAY support subscribing to all command types by accepting requests at `/<endpoint_token>/command` resource path. In that case, command representation MUST include `type` property that is a string identifying the command type.
+
+### Response
+The server MUST respond with all outstanding commands of the specified command type.
+
+The response MUST be a UTF-8 encoded array with the schema defined in [command-response.schema.json](./command-response.schema.json) file.
+
+```json
+{
+  "$schema": "http://json-schema.org/schema#",
+  "title": "11/CEP command response schema",
+
+  "type": "array",
+  "items": {
+    "type": "object",
+    "properties": {
+      "id": {
+        "type": "integer",
+        "description": "ID of the command."
+      },
+      "payload": {
+        "description": "The command payload to be interpreted by the endpoint."
+      }
+    },
+    "required": [
+      "id"
+    ],
+    "additionalProperties": false
+  }
+}
+```
```

Comment:
The message is identified by the id as defined per 1/KP.

The command execution order is no defined, so when a server sends an array of commands to a client, it may select and execute any of them and in any order. The `id` field of the command is used to match command results back to the command.